src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.cpp

```diff
@@ -19,6 +19,24 @@ AbstractASTNode::AbstractASTNode(ASTNodeType node_type) : _type(node_type) {}
 
 bool AbstractASTNode::is_optimizable() const { return true; }
 
+bool AbstractASTNode::subtree_is_read_only() const {
```

Comment:
`subtree_is_read_only()` and `subtree_is_validated()` are perfect cases for functions that should be implemented outside of AST node classes, but in the ast_utils module as proposed here #446 - they can operate on the public interface of the AST node types - and their logic doesn't need to be spread over multiple classes!

Reply:
Also, I think you should doc/exemplify what these functions are there for